pkg/kv/kvserver/replica_init.go

```diff
@@ -173,28 +185,26 @@ func (r *Replica) setStartKeyLocked(startKey roachpb.RKey) {
 	r.startKey = startKey
 }
 
-// loadRaftMuLockedReplicaMuLocked will load the state of the replica from disk.
-// If desc is initialized, the Replica will be initialized when this method
-// returns. An initialized Replica may not be reloaded. If this method is called
-// with an uninitialized desc it may be called again later with an initialized
-// desc.
+// loadRaftMuLockedReplicaMuLocked loads the state of the initialized replica
+// from storage. After this method returns, Replica is initialized, and can not
+// be loaded again.
 //
-// This method is called in three places:
+// This method is called in two places:
 //
 //  1. newReplica - used when the store is initializing and during testing
-//  2. tryGetOrCreateReplica - see newUnloadedReplica
-//  3. splitPostApply - this call initializes a previously uninitialized Replica.
+//  2. splitPostApply - this call initializes a previously uninitialized Replica.
 func (r *Replica) loadRaftMuLockedReplicaMuLocked(desc *roachpb.RangeDescriptor) error {
 	ctx := r.AnnotateCtx(context.TODO())
+	if !desc.IsInitialized() {
+		log.Fatalf(ctx, "r%d: cannot load an uninitialized replica", desc.RangeID)
```

Comment:
Since we're here - we're fataling here (and below) but we have the option to return an error, so we should prefer doing that via `return errors.AssertionFailedf(...)`.

Reply:
Done, but I'm not sure - is that a "canonical" approach? The good thing about `Fatalf` is that it prints the stack which is somewhat useful to understand the problem. If we return the error all the way up the stack, this information can be lost.

Reply:
That's true, but the error does contain the stack, and it's printed at least when it causes an assertion failure higher up.